### PR TITLE
JAMES-2343 Add buffered DeferredFileOutputStream and use it in MimeMe…

### PR DESCRIPTION
**Pull request by chibenwa**
…ssageInputStreamSource

From the original work of @chikei

Opened as the repository used for https://github.com/apache/james-project/pull/105 was (apparently?) deleted...

### CODE REVIEW

**Comment by chibenwa** [comment]
## Before

Capture d’écran de 2020-06-18 09-08-50

## After

Capture d’écran de 2020-06-18 09-08-50

## Conclusions:

 - Twice the rate, which then impact other metrics (p99 & errors)
 - x2 improvement for p50, normal processing time improved. ~ 2 mean time reduction reflects that too.

Exceptions are due to ObjectStorage (signle node in this test setup) being overwhelmed:

```
 org.jclouds.http.HttpResponseException: Server rejected operation connecting to PUT http://swift:8080/v1/TMP_test/default/7d69559f2b230319d17101a92802d4bb13639e2c23bf61e7b1ec7378b59c8669 HTTP/1.1
```

**Comment by chibenwa** [comment]
Build is :green_apple: , let's wait for reviews :-)

File: server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+     * @param threshold  The number of bytes at which to trigger an event.
+     * @param prefix Prefix to use for the temporary file.
+     * @param suffix Suffix to use for the temporary file.
+     * @param directory Temporary file directory.
+     */
+    public BufferedDeferredFileOutputStream(final int threshold, final String prefix, final String suffix, final File directory) {
+        this(threshold, null, prefix, suffix, directory);
+        if (prefix == null) {
+            throw new IllegalArgumentException("Temporary file prefix is missing");
+        }
```
Comment (blackheaven) [inline server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java:117]: Preconditions

File: server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+     * Switches the underlying output stream from a memory based stream to one
+     * that is backed by disk. This is the point at which we realise that too
+     * much data is being written to keep in memory, so we elect to switch to
+     * disk-based storage.
+     *
+     * @exception IOException if an error occurs.
+     */
+    @Override
+    protected void thresholdReached() throws IOException {
+        if (prefix != null) {
```
Comment (blackheaven) [inline server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java:165]: `Optional`

File: server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            outputFile = File.createTempFile(prefix, suffix, directory);
+        }
+        final FileOutputStream fos = new FileOutputStream(outputFile);
+        try {
+            memoryOutputStream.writeTo(fos);
+        } catch (IOException e) {
+            fos.close();
+            throw e;
+        }
```
Comment (blackheaven) [inline server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java:174]: `finally`

File: server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        currentOutputStream = new BufferedOutputStream(fos, getThreshold());
+        memoryOutputStream = null;
```
Comment (blackheaven) [inline server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java:176]: `Optional`

File: server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+     * Writes the data from this output stream to the specified output stream,
+     * after it has been closed.
+     *
+     * @param out output stream to write to.
+     * @exception IOException if this stream is not yet closed or an error occurs.
+     */
+    public void writeTo(final OutputStream out) throws IOException {
+        // we may only need to check if this is closed if we are working with a file
+        // but we should force the habit of closing wether we are working with
+        // a file or memory.
```
Comment (blackheaven) [inline server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java:248]: the strategy pattern should be used then

File: server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+        final BufferedDeferredFileOutputStream dfos =
+            new BufferedDeferredFileOutputStream(testBytes.length - 5,  testFile);
+        try
+        {
+            dfos.write(testBytes, 0, testBytes.length);
+            dfos.close();
+        }
+        catch (final IOException e) {
+            fail("Unexpected IOException");
+        }
```
Comment (blackheaven) [inline server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java:121]: Rely on the test case `Exception`

File: server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+        {
+            dfos.write(testBytes, 0, chunkSize);
+            dfos.write(testBytes, chunkSize, chunkSize);
+            dfos.write(testBytes, chunkSize * 2,
+                testBytes.length - chunkSize * 2);
+            dfos.close();
+        }
+        catch (final IOException e) {
+            fail("Unexpected IOException");
+        }
```
Comment (blackheaven) [inline server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java:157]: idem

File: server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+
+            assertFalse(testFile.exists());
+            assertTrue(dfos.isInMemory());
+
+            try {
+                dfos.writeTo(baos);
+                fail("Should not have been able to write before closing");
+            } catch (final IOException ioe) {
+                // ok, as expected
+            }
```
Comment (blackheaven) [inline server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java:191]: `assertThrownBy`

File: server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java (hunk 1 above)
Comment (chibenwa) [reply server/container/core/src/main/java/org/apache/james/server/core/BufferedDeferredFileOutputStream.java:117]: The BufferedDefferedFileOutputStream is (mostly) taken from commons-io, I would like to get that code as close as possible from the original version, if you don't mind.

File: server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java (hunk 8 above)
Comment (chibenwa) [reply server/container/core/src/test/java/org/apache/james/server/core/BufferedDeferredFileOutputStreamTest.java:191]: idem for the tests

**Comment by Arsnael** [comment]
> Twice the rate, which then impact other metrics (p99 & errors)

Why do we get twice the rate?

**Comment by chibenwa** [comment]
Because it is faster. Buffering things before writing them to a file is a nice idea...

**Comment by mbaechler** [comment]
Please always describe the load testing scenario you run and why.
Here you should run not only for big mails but also for small ones as they should be kept in memory under a given threshold.

Regarding the analysis, you should try to inject only half the current load (or a third) so that object storage is not a problem anymore and you can measure the latency anyway. WDYT?

**Comment by Arsnael** [comment]
> Because it is faster. Buffering things before writing them to a file is a nice idea...

Ok same amount of time, so it means it's faster. Was not sure the schenario was like that :)

**Comment by chibenwa** [comment]
> Please always describe the load testing scenario you run and why.

SmtpMessage sending, big size.

The idea is to exercise enqueue speed, what the original contributor did tested in isolation.

Here is the link to the scenario: https://github.com/chibenwa/james-gatling/blob/smtp-benchs/src/test/scala-2.12/org/apache/james/gatling/simulation/smtp/SmtpNoAuthenticationNoEncryptionBigBodySimulation.scala (slightly adapted)

> Here you should run not only for big mails but also for small ones as they should be kept in memory under a given threshold.

> Regarding the analysis, you should try to inject only half the current load (or a third) so that object storage is not a problem anymore and you can measure the latency anyway. WDYT?

Let's run that then...

**Comment by chibenwa** [comment]
I did launch the 4 tests (with a reduced user count) let's see the output on monday

**Comment by chibenwa** [comment]
```
Simultation: org.apache.james.gatling.simulation.smtp.SmtpNoAuthenticationNoEncryptionSimulation
chibenwa/james-gatling master branch (btw could be nice to specify custom branches here)
```

## Before

Capture d’écran de 2020-06-21 11-29-00

## After

Capture d’écran de 2020-06-21 11-29-53


## Conclusion for small mail reception

 - Less errors but slightly slower

Which is good!

Regarding a lower rate for big mails, one of the simultation aborted, I relaunch it.

**Comment by chibenwa** [comment]
Here is the result of the run with less throughtput and big messages:

## master

Capture d’écran de 2020-06-22 10-46-56

## This changeset

Capture d’écran de 2020-06-22 10-48-18

**Comment by mbaechler** [comment]
We really need small messages too

**Comment by chibenwa** [comment]
> We really need small messages too

https://github.com/apache/james-project/pull/211#issuecomment-647077562

**Comment by mbaechler** [comment]
> ```
> Simultation: org.apache.james.gatling.simulation.smtp.SmtpNoAuthenticationNoEncryptionSimulation
> chibenwa/james-gatling master branch (btw could be nice to specify custom branches here)
> ```
> 
> ## Before
> 
> Capture d’écran de 2020-06-21 11-29-00
> ## After
> 
> Capture d’écran de 2020-06-21 11-29-53
> ## Conclusion for small mail reception
> 
>     * Less errors but slightly slower
> 
> 
> Which is good!
> 
> Regarding a lower rate for big mails, one of the simultation aborted, I relaunch it.

16% slow down is a lot, don't you think ?

**Comment by chibenwa** [comment]
> 16% slow down is a lot, don't you think ?

That's a very specific workload, with only messages below 1 KB that don't represent a real production usage.

**Comment by chibenwa** [comment]
I will rerun this

**Comment by chibenwa** [comment]
Shall I take time rebase this work, or can I just discard it?

**Comment by chibenwa** [comment]
I would actually REALLY be interested to know "how this plays" with https://github.com/apache/james-project/pull/309

I'll run benches <3

**Comment by chibenwa** [comment]
We ended up merging this work.

Thanks for your time and the contribution, it ended up not being lost :-)

**Pull request closed without merging**